### ISSUE TITLE
Pagination Problem

### ISSUE THREAD

**Issue by ngonzalezromero** [JasperFx/marten#480]
Hi.
I have 11 registers in database

With this code

```
   public async Task<PagingResult<Message>> PagingAsync(Expression<Func<Message, bool>> condition, string columnOrder, bool orderDescending, int pageIndex, int pageSize)
    {
        var session = _documentStore.OpenSession();
        PagingResult<Message> data = new PagingResult<Message>();
        IList<Message> list;

        try
        {
            QueryStatistics stats = null;
            var order = _apiUtil.GetLambda<Message>(columnOrder);

            var query = session.Query<Message>().Stats(out stats).Where(condition);

            if (orderDescending)
            {
                list = await query.OrderByDescending(order).Skip(pageIndex).Take(pageSize).ToListAsync();
            }
            else
            {
                list = await query.OrderBy(order).Skip(pageIndex).Take(pageSize).ToListAsync();
            }

            if (list != null && list.Any())
            {
                data.List = list;
                data.Count = stats.TotalResults;
            }
        }
        catch (Exception ex)
        {
            _logger.LogCritical(_apiUtil.GetExceptionDetails(ex));
        }
        finally
        {
            session.Dispose();
        }
        return data;
    }
```

if i send  pageIndex =1 and  pageSize =10, Marten return 10 registers.
But when pageIndex =2 and  pageSize =10, Marten return 9 register, but this is no correct, because should be return only 1( the page 2 of 11 element only contains 1 element)

cc: @jeremydmiller

**Comment by jeremydmiller** [comment]
I think you need something like:

```
Skip((pageIndex - 1) * pageSize);
```

The Skip() uses the total number of items to skip. It doesn't have any logic around paging per se.

**Comment by ngonzalezromero** [comment]
Thank you @jeremydmiller 
Nice project !!!
Regards

**Comment by jeremydmiller** [comment]
@ngonzalezromero If you're bored and need a little OSS side thing, it wouldn't hurt to have a better paging model in Marten where you could do something more like:

```
session.Query<Document>().Page(page #, page size);
```

but just have it delegate to Skip() & Take() internally.

**Comment by ngonzalezromero** [comment]
Uf.. It would be nice !!

if you want.. I could do it ..
Just please give me some instruction where I must add the code and then I will send the PR..

I would love to contribute to this project

**Comment by mdissel** [comment]
And optionally return a PagedResult with recordcount, numberofpages, etc.

`PagedResult  result = session.Query<Document>().PagedResult(page, pagesize);`

**Comment by jeremydmiller** [comment]
@ngonzalezromero Honestly, I was just thinking an extension method off of IQueryable<T> that delegated to Skip() and Take() behind the scenes. There's already an existing `QueryableExtensions` class for that. That wouldn't work OOTB for compiled queries, but one of us could add the Linq customizations to match later if you don't wanna have to get into Relinq mechanics.

@mdissel What would your PagedResult look like?

**Comment by mdissel** [comment]
Something like

```
public class PagedResult<T>
{
    public IList<T> Results { get; set; }
    public int CurrentPage { get; set; }
    public int PageCount { get; set; }
    public int PageSize { get; set; }
    public int RowCount { get; set; }
}
```

**Comment by ngonzalezromero** [comment]
I can do it!!

**Comment by khalidabuhakmeh** [comment]
Since Marten implements `IQueryable` you should be able to use the `PagedList` nuget package.

**Comment by jeremydmiller** [comment]
@khalidabuhakmeh From what I'm seeing looking through the PagedList codebase, I don't think that's going to work the way you'd think. That would force you to do the paging on the client, wouldn't it?

Either way, I see this as just some syntactic sugar over what's already there. It's going to be more work to make this happen w/ compiled queries, but otherwise shouldn't be very hard.

**Comment by khalidabuhakmeh** [comment]
@jeremydmiller we use it successfully with both Entity Framework and even RavenDB. The only down side is it does two queries. 1 to get the total count then to use that to do the paging.

**Comment by jeremydmiller** [comment]
Hey all, I'm going to let this one slide to 1.3 for now.

**Comment by jeremydmiller** [comment]
I'm going to call this one closed by #639
